### DISCUSSION TITLE
Umi 打包如何将配置文件单独分离出来

### DISCUSSION THREAD

**Discussion by ByYogi** [umijs/umi#11136] [Q&A]
使用`build`打包出来的文件中。定义的api url很难找到。我想在测试环境配置一个api，生产环境另一个api。
目前就是只能在源代码改一个url再打包，再改再打包。每次测试都很麻烦。

请问各位大佬有没有相关解决方案，将配置单独抽离。这样就只需要打包一次，不同环境手动改一下配置文件url就行了

**Comment by fz6m** [comment]
弄一个环境变量，在不同环境下启动有不同值，然后在 `define` 注入到项目里面，判断下这个变量就可以区分环境。

**Reply by ByYogi** [reply]
要的是打包后手动配置，而不是构建时读取配置

**Reply by fz6m** [reply]
设定一个仅在线下环境生效的逻辑，从 localStorage 里或者 url 的 query 上获取你想要动态改变的值。

**Reply by ByYogi** [reply]
这么麻烦吗？就不能抽离一个配置文件读取吗

**Reply by licc12** [reply]
实现方案：
1.代码中添加config文件，改文件中放入的你的配置文件config.js；
2..umirc.ts文件添加如下配置：
```js

  copy: [
    {
      from: 'src/config/config.js',
      to: 'config/config.ts',
    },
  ],
  headScripts: [{ src: `/你的publicPath的值/config/config.js`, async: true }], // async确保配置文件在项目中先被加载
  
```

结果：
build 打包后：
1.dist文件夹中会出现一个config文件夹，里面是拷贝出来的config.js配置文件；
2.如果部署在镜像中，可以在镜像中挂载config目录；

**Reply by licc12** [reply]
3.index.html文件头中会有一个<script src="/你的publicPath的值/config/config.js">

**Reply by ByYogi** [reply]
但是，生成打包默认用的是 config/config.ts 的配置文件。而且我用了{hash: true} 哈希打包。没有文件夹分类。就是加了head script也不会用这个指定的配置文件js。而是用打包好的名称是hash生成的js

**Reply by licc12** [reply]
src/config/config.js 这个文件夹是你自己新建的，你可以写到其它目录下。整体思路不是用umi本身的config.js文件，而是自己写一个改文件，本地开发时从改文件中引入；发布部署时将改文件通过copy拷贝到dist目录下，
```js
  copy: [
    {
      from: 'src/config/config.js',
      to: 'config/config.ts',
    },
  ],
```
这样你如果用了docker镜像打包，就可以将dist下的config目录进行挂载操作，从而实现打包后还可以外部动态配置

**Reply by ByYogi** [reply]
copy: [
    {
      from: 'src/config/config.js',
      to: 'config/config.ts',
    },
  ],
这个配置不也是config.js里面的吗。要怎么弄啊。不用umi本身的config.js文件也要设置一个配置文件路径啊。设置配置文件的路径也是在配置文件里设置吗。有没有代码呀大佬
